activerecord/lib/active_record/encryption/encryptable_record.rb

```diff
@@ -121,12 +120,22 @@ def override_accessors_to_preserve_original(name, original_attribute_name)
             end)
           end
 
+          def load_schema!
+            super
+
+            add_length_validation_for_encrypted_columns if ActiveRecord::Encryption.config.validate_column_size
+          end
+
+          def add_length_validation_for_encrypted_columns
+            encrypted_attributes&.each do |attribute_name|
+              validate_column_size attribute_name
+            end
+          end
+
           def validate_column_size(attribute_name)
             if table_exists? && limit = columns_hash[attribute_name.to_s]&.limit
```

Comment:
This is slightly outside the diff and apologies for the nitpick comment.

I'm wondering why we call `table_exists?` for every `attribute_name` is the table for every attribute the same? Can we move this call a little higher and return early in the method `add_length_validation_for_encrypted_columns`?

Reply:
(found this whilst reading the code, it's totally unrelated to the changes so I'm not sure if it's worth doing) I think we can move this https://github.com/rails/rails/pull/43009/files#R51 slightly higher too, because the `scheme` looks to be the same for every attribute.

Reply:
> I'm wondering why we call table_exists? for every attribute_name is the table for every attribute the same? 

Great catch @hahmed. With the new approach where we only add validations when the schema loads, so this check is not needed. I had added it back in the day to deal with the situation where Rails tried to load a class with encrypted attributes without a table (such as when you don't use action text but Rails loads it). I removed the check 👍.

Reply:
>  I think we can move this https://github.com/rails/rails/pull/43009/files#R51 slightly higher too, because the scheme looks to be the same for every attribute.

@hahmed sorry which line is that? I can't see which piece of code you are referring to in that link 😅. 

Thanks for the review!

Reply:
Whoops, I meant this line here https://github.com/basecamp/rails/blob/f0fe547e20476e00637d15de57c14532fec6c01c/activerecord/lib/active_record/encryption/encryptable_record.rb#L51

So I think that we could do something like this?

```
def encrypts(*names, key_provider: nil, key: nil, deterministic: false, downcase: false, ignore_case: false, previous: [], **context_properties)
          self.encrypted_attributes ||= Set.new # not using :default because the instance would be shared across classes
          scheme = scheme_for key_provider: key_provider, key: key, deterministic: deterministic, downcase: downcase, \
              ignore_case: ignore_case, previous: previous, **context_properties

          names.each do |name|
            encrypt_attribute name, scheme
          end
        end
```

As in, we don't need to compute the scheme for every attribute?

Reply:
Oh totally @hahmed, good eye on that one 🙏! I just pushed the suggested change.

Reply:
Created a new PR with the two suggestions https://github.com/rails/rails/pull/43021

Reply:
(added you as coauthor there)